Docker-Compose for Development

kierachell opened the pull request:
connects https://github.com/department-of-veterans-affairs/caseflow/issues/2581

- Updated Dockerfile to streamline image building
- There's two `Procfile`s now - one for Docker and one for regular work
- Run `npm install` in `client/` directory to get your `node_modules`

Then:
- Docker-compose with hot module reloading setup:
```
docker-compose build
docker-compose run rails bundle install
docker-compose run rails rake db:create
docker-compose run rails rake db:schema:load
docker-compose up
```
When restarting the server/compose, you will have to delete the `server.pid` in `tmp/pids/` folder
- Outstanding issue: docker containers write files as `root` to host :(
- I tested this on Linux. Docker is quite different on Mac so please let me know what's different


To run tests I added 
`docker_test_runner.sh` which can be used with:
```
docker-compose run rails ./docker_test_runner.sh
```

![it works](https://user-images.githubusercontent.com/18075411/28294337-77c73538-6b27-11e7-8d10-6436b3919ade.gif)

- Updated Jenkinsfile as well to make frontend tests pass


## Questions
- this whole webpack server binding is a bit of `magic` for me so let me know if that needs to be changed
- commented out lines in Dockerfile are just an aspiration that one day mounted volumes will not be written as `root`
- Add option to not mount PG and Redis data as it requires `root` to remove
- Not sure how to deal with `tmp/pids` thing either - also requires `root` to kill

askldjd commented on Procfile:
```diff
@@ -1,2 +1,2 @@
-rails: REACT_ON_RAILS_ENV=HOT bundle exec rails s -p 3000
+rails: REACT_ON_RAILS_ENV=HOT bundle exec rails s -b '0.0.0.0' -p 3000
```
I would avoid binding all interfaces for security reasons. For docker, this is okay. Outside of docker, definitely not a good default.

mogrenAtWork commented on Dockerfile:
```diff
@@ -37,46 +46,65 @@ RUN apt-get update -qq && apt-get install -qq -y \
+WORKDIR /caseflow/client
+RUN npm install
+
+WORKDIR /caseflow
```
Is there a need to specify `WORKDIR` so many times?

mogrenAtWork commented on docker-compose.yml:
```diff
@@ -13,34 +14,38 @@ services:
-    build: .
+
+  rails:
+    image: caseflow
```
This won't work for people who don't have access to the DSVA docker repo. I would just rely on the local Dockerfile and your local Docker repo. I know it sucks having to build it, but we should keep everything self-contained

kierachell replied:
So the idea was that you can run `docker build -t caseflow .` (as I commented in the PR) locally to build your own Dockerfile image and that would tag it `caseflow` so then compose would use your own

mogrenAtWork replied:
I like this sequence of commands to get up and running:
```
docker-compose build
docker-compose run rails rake db:create
docker-compose up
```
That way you don't have to worry about any funniness if you don't use `docker build -t caseflow`.

joofsh replied:
My 2 cents: another benefit of running `docket-compose build` each time is to ensure that the build steps stay updated/correct.

mogrenAtWork commented on Dockerfile:
```diff
@@ -37,46 +46,65 @@ RUN apt-get update -qq && apt-get install -qq -y \
+	--fix-missing \
+	&& rm -rf /var/lib/apt/lists/*
+
+# ARG user=caseflow
```
Delete commented out code

mogrenAtWork commented on config/environments/development.rb:
```diff
@@ -27,6 +27,11 @@
 
+  # Docker configurations for hot module reloading
+  # Also stops web-console spam when trying to get webfonts
+  config.web_console.whitelisted_ips = ['172.18.0.0/12', '192.168.0.0/16']
```
I'm not sure what address `172.18.0.0` but this seems to work without it

kierachell replied:
This is the error I am seeing without that configuration:
```
[2017-07-19 00:04:24 +0000] Started GET "/assets/react/_comment.self-362b082cedc0425bd6abfd1e7353fbab8fe672af97a0f592b58147792b090662.css?body=1" for 172.18.0.1 at 2017-07-19 00:04:24 +0000
rails_1  | 00:04:24 rails.1  | [localhost] [78e459f7-ec72-4c55-b79a-120f65d4b09b] [localhost] [Fake User america@example.com] Cannot render console from 172.18.0.1! Allowed networks: 192.168.0.0/192.168.255.255, 127.0.0.0/127.255.255.255, ::1
rails_1  | 00:04:24 rails.1  | [localhost] [78e459f7-ec72-4c55-b79a-120f65d4b09b] [localhost] [Fake User america@example.com] [2017-07-19 00:04:24 +0000] Cannot render console from 172.18.0.1! Allowed networks: 192.168.0.0/192.168.255.255, 127.0.0.0/127.255.255.255, ::1
```
Which indcates that 172* space is used by the docker to communicate with host and it cannot render JS assets because the IP is from an unexpected space (not local)

mogrenAtWork commented on config/environments/development.rb:
```diff
@@ -27,6 +27,11 @@
+  # Docker configurations for hot module reloading
+  # Also stops web-console spam when trying to get webfonts
+  config.web_console.whitelisted_ips = ['172.18.0.0/12', '192.168.0.0/16']
+  # config.webpack_hot = true
```
Remove this commented out line

mogrenAtWork commented on docker-compose.yml:
```diff
@@ -3,6 +3,7 @@ services:
   db:
     image: postgres:9.3.14
     environment:
+
```
Remove extra new line

mogrenAtWork commented on client/webpack-hot-server.js:
```diff
@@ -32,7 +32,7 @@ const devServer = new WebpackDevServer(compiler, {
   }
 });
 
-devServer.listen(hotRailsPort, 'localhost', (err) => {
```
Is there any reason to change this? It seems to work with `localhost`

mogrenAtWork commented on client/webpack.hot.config.js:
```diff
@@ -21,7 +21,7 @@ config.entry.push(
 config.output = {
   filename: 'webpack-bundle.js',
   path: path.join(__dirname, '../app/assets/webpack'),
-  publicPath: `http://localhost:${hotRailsPort}/`
```
Same here

kierachell replied:
I am seeing this error without explicit 0.0.0.0
Failed to load resource: net::ERR_CONNECTION_REFUSED
http://localhost:3500/webpack-bundle.js

mogrenAtWork commented on config/database.yml:
```diff
@@ -14,32 +14,43 @@ default: &default
+
+
+production:
+  <<: *default
```
Looks like the values added for the default Postgres connection will affect `prod`, even if the `POSGRES_URL` is set with the username and password, will this be okay?

kierachell replied:
This is just moved over to order settings by db -> env as opposed to env -> db

I do _think_ there's any functional change

mogrenAtWork replied:
Oh nevermind, I see you didn't alter default at all

mogrenAtWork commented on docker-compose.yml:
```diff
@@ -13,34 +14,38 @@ services:
+    command: foreman start -f Procfile.docker
     ports:
       - "3000:3000"
+      - "3500:3500"
```
Is there a need to expose port 3500? I think only the Docker container would need access to it, but I could be wrong

kierachell commented:
- @mogrenAtWork Addressed most of the comments; let me know if you have more questions
- Added `docker_test_run.sh` for good measure
- Could use some more feedback from @NickHeiner @sharonwarner and other JS-loving engineers

mogrenAtWork commented on Jenkinsfile.pr:
```diff
@@ -96,7 +97,6 @@ podTemplate(cloud:'minikube', label:'caseflow', containers: [
             container('caseflow-test-runner') {
```
Delete this stage if Xvfb is still running from the previous stage

kierachell replied:
I am literally testing it right now if it works :D

mogrenAtWork commented on Dockerfile:
```diff
@@ -37,46 +36,55 @@ RUN apt-get update -qq && apt-get install -qq -y \
+RUN bundle install --without production staging
+
+# Entrypoint for the container
+CMD ["foreman", "start", "-f", "Procfile.docker"]
```
I would change this to `ENTRYPOINT ["/bin/bash"]` since docker-compose starts up foreman. If you execute this without docker-compose, it will fail

mogrenAtWork commented on config/database.yml:
```diff
@@ -14,32 +14,43 @@ default: &default
 
+demo:
+  <<: *default
+  username: <%= `whoami` %>
```
You can get rid of the `username: <%= `whoami` %>` now that it's in default

joofsh commented on config/database.yml:
```diff
@@ -14,32 +14,41 @@ default: &default
+  host: <%= ENV['POSTGRES_HOST'] %>
+  username: <%= ENV['POSTGRES_USER'] || `whoami` %>
+  # If we set the value to nil when the variable isn't present,
+  # the k8s Jenkins environment won't connect
```
sorry I'm a bit out of the loop here. Why do we need to do this conditional? And if we do need it, should we do this only in the `test` block?

mogrenAtWork replied:
The conditional only needs to be in the `test` block. I added it for the Jenkins test environment because the default setup required a password, Travis does not. 

There are different behaviours when the connection string as an empty password vs no password at all and this conditional was added to not disrupt the current Travis setup.

joofsh commented on docker_test_runner.sh:
```diff
@@ -0,0 +1,17 @@
+rake db:create
+rake db:schema:load
+
+rake lint
```
I'm fine in general making changes to the rake test commands, but for now these should mirror what we run in travis, no? That or make the corresponding changes to the travis yml

kierachell replied:
This is not for travis, this is to deprecate travis and mirror Jenkins setup (aside from parallel execution)

joofsh commented on Procfile.docker:
```diff
@@ -0,0 +1,2 @@
+rails: REACT_ON_RAILS_ENV=HOT bundle exec rails s -b '0.0.0.0' -p 3000
+client: cd client && npm run dev:hot
```
Similar to @mogrenAtWork feedback below, do we actually need this separate `Procfile.docker`? The only different seems to be the `-b 0.0.0.0`, and that still works fine for me locally. I tested running `foreman start -f Procfile.docker` locally (outside docker) and it worked fine

kierachell replied:
I wanted to keep configurations separate between host and docker env, so I just added a separate file instead of changing the current setup too much.
All this stuff is not meant to replace current functionality, only provide options.

joofsh commented:
Overall feedback: I was successfully able to run this branch locally using docker. But it was *significantly* slower than running it natively on my mac. Based on the current speed, this is not a viable development option.

<img width="1667" alt="screen shot 2017-07-18 at 11 48 13 pm" src="https://user-images.githubusercontent.com/2256237/28349950-ca3d277c-6c13-11e7-9385-6f9ef6727510.png">
<img width="1671" alt="screen shot 2017-07-18 at 11 48 43 pm" src="https://user-images.githubusercontent.com/2256237/28349951-ca3d308c-6c13-11e7-9d77-e24965490f21.png">

mogrenAtWork commented on Dockerfile:
```diff
@@ -37,46 +36,55 @@ RUN apt-get update -qq && apt-get install -qq -y \
-RUN gem install bundler
+ENV PATH=~/.rbenv/plugins/ruby-build/bin:$PATH
+
+RUN rbenv install -v 2.2.4 && \
```
Agreed, if it fails in the middle of this command, Docker would have to roll back any previous changes. 

I don't see the added benefit of chaining these commands together

kierachell replied:
https://docs.docker.com/engine/userguide/eng-image/dockerfile_best-practices/#minimize-the-number-of-layers

If you prefer more layers, I can do that.

mogrenAtWork commented:
I'm experiencing similar slowdowns, if not worse, as @joofsh on my Macbook Pro. I don't think this is  a viable dev environment at this speed. Docker also eats up A LOT of disk space for each of the images.

Let's get it to a stable environment that you can execute the tests in and not worry about making it a complete development environment.

kierachell commented:
12 seconds sounds about right on initial load without Docker depending on the weather:
```
08:46:33 rails.1  | [localhost] [c272d325-bb04-46d3-895a-5c329e7bf067] [localhost] Started GET "/test/users"
```

and after a ton of spam:
```
08:52:53 rails.1  | [localhost] [182fdc1a-37fb-456d-b82f-02e7f52411e0] [localhost] [Fake User america@example.com] [2017-07-19 08:52:53 -0400] Started GET "/as
```
